src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -926,77 +986,67 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
             finally
             {
                 _headerBuffer.Discard(_headerBuffer.ActiveMemory.Length);
-                _writerLock.Release();
+                _headerSerializationLock.Release();
             }
 
             return http2Stream;
         }
 
-        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer)
+        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
         {
             ReadOnlyMemory<byte> remaining = buffer;
 
             while (remaining.Length > 0)
             {
                 int frameSize = Math.Min(remaining.Length, FrameHeader.MaxLength);
 
-                frameSize = await _connectionWindow.RequestCreditAsync(frameSize).ConfigureAwait(false);
+                // Once credit had been granted, we want to actually consume those bytes.
+                frameSize = await _connectionWindow.RequestCreditAsync(frameSize, cancellationToken).ConfigureAwait(false);
 
                 ReadOnlyMemory<byte> current;
                 (current, remaining) = SplitBuffer(remaining, frameSize);
 
-                await AcquireWriteLockAsync().ConfigureAwait(false);
+                // It's possible that a cancellation will occur while we wait for the write lock. In that case, we need to
+                // return the credit that we have acquired and don't plan to use.
                 try
                 {
-                    _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
-                    WriteFrameHeader(new FrameHeader(current.Length, FrameType.Data, FrameFlags.None, streamId));
-                    current.CopyTo(_outgoingBuffer.AvailableMemory);
-                    _outgoingBuffer.Commit(current.Length);
-
-                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                    await StartWriteAsync(FrameHeader.Size + current.Length, cancellationToken).ConfigureAwait(false);
                 }
-                finally
+                catch (OperationCanceledException)
                 {
-                    ReleaseWriteLock();
+                    _connectionWindow.AdjustCredit(frameSize);
+                    throw;
                 }
+
+                WriteFrameHeader(new FrameHeader(current.Length, FrameType.Data, FrameFlags.None, streamId));
+                current.CopyTo(_outgoingBuffer.AvailableMemory);
+                _outgoingBuffer.Commit(current.Length);
+
+                await FinishWriteAsync(false, cancellationToken).ConfigureAwait(false);
             }
         }
 
         private async ValueTask SendEndStreamAsync(int streamId)
```

Comment:
ValueTask => Task

For things that don't affect public surface area, for methods implemented with async/await, default to using:
- `Task` if the method doesn't return a value
- `ValueTask<T>` if the method does return a value

Reply:
I've gone ahead and updated all of the instances you commented on, and a couple of other locations based on what you suggested here. Thanks.